fixes for Processing geometry checker algorithms

alexbruy opened the pull request:
## Description

Various fixes for Processing check/fix geometry algorithms:

- raise an exception in geometry checker algorithms if selected unique ID field contains duplicated values to prevent failures at the geometry repair step (fixes #61398)
- avoid use of `QgsProject::instance()` in algorithms
- correctly mark algorithms which require project with the corresponding flag
- added cancellation support in geometry checks

Djedouas reviewed (approved):
Looks good to me, thanks!

alexbruy commented on src/analysis/processing/qgsalgorithmcheckgeometryangle.cpp:
```diff
@@ -136,6 +136,12 @@ QVariantMap QgsGeometryCheckAngleAlgorithm::processAlgorithm( const QVariantMap
     throw QgsProcessingException( QObject::tr( "Missing field %1 in input layer" ).arg( uniqueIdFieldName ) );
 
+  // ensure that unique ID field is indeed unique
+  if ( input->featureCount() != input->uniqueValues( uniqueIdFieldIdx ).count() )
```
Probably yes, from the quick look at the code this can be done either when building a feature pool or when running a check itself. Probably @Djedouas can provide more feedback on this matter.

Djedouas replied:
The iteration on the features is done inside the geometry checker API, and not in the processing itself.

In my opinion, trying to give the feature pool awareness of a unique field, to verify inside, breaks some abstraction. It is the processing that requires a unique field to be truly unique, not the geometry checker or the feature pool.

That's why I originally chose to check the unique field in the fix algorithms, which iterates on the errors to fix.

I would say that the method in this PR is the correct method, but it is a compromise with performance on very big layers. And big layers are what we expect to have with these check algorithms.

Not sure what to recommend here.

alexbruy committed:
optionally check for duplicated unique IDs when performing geometry
checks

If the uniqueIdFieldIndex in QgsGeometryCheckerContext is a valid field
index, check that there are no duplicated values in that field and fail
if this is the case.

alexbruy committed:
add tests

alexbruy committed:
add cancellation support for collecting errors in geometry checks

alexbruy committed:
more const

alexbruy commented:
@nyalldawson, @Djedouas I have updated PR with another approach. Now check for duplicated IDs is performed in the geometry checker when unique ID field index is set in the geometry checker context.

I also added support for cancellation as most of the checks do not have it.

@Djedouas may I ask why are you using `QgsProject::instance()` in the algorithms and not project from the context? Also I noticed that all algorithms are marked with `NoThreading` flag. According to you

>  And big layers are what we expect to have with these check algorithms.

But with checks/fixes running in the main thread they seem not very usable with large datasets.

alexbruy committed:
add cancellation support in more places

alexbruy committed:
mark geometry checker algorithms as requiring project

alexbruy committed:
do not use QgsProject::instance() in the majority of the geometry
checker algorithms

The small gaps check still rely on QgsProject::instance() as well
as corresponding fixer.

alexbruy committed:
make small gaps check/fix work without QgsProject::instance()

nyalldawson commented on src/analysis/processing/qgsalgorithmcheckgeometryarea.cpp:
```diff
@@ -63,7 +63,7 @@ QString QgsGeometryCheckAreaAlgorithm::shortHelpString() const
 Qgis::ProcessingAlgorithmFlags QgsGeometryCheckAreaAlgorithm::flags() const
 {
-  return QgsProcessingAlgorithm::flags() | Qgis::ProcessingAlgorithmFlag::NoThreading;
+  return QgsProcessingAlgorithm::flags() | Qgis::ProcessingAlgorithmFlag::RequiresProject;
```
I don't think these algorithms ARE thread safe -- at least, there's access to the project from the process step, which hints that there's at least some logic which would need to be moved to the prepare stage. Can we leave the no thread flag for now and then address that in a follow up?

nyalldawson commented on src/analysis/vector/geometry_checker/qgsgeometrycheckcontext.h:
```diff
@@ -34,7 +34,7 @@ class ANALYSIS_EXPORT QgsGeometryCheckContext
      * Creates a new QgsGeometryCheckContext.
      */
-    QgsGeometryCheckContext( int precision, const QgsCoordinateReferenceSystem &mapCrs, const QgsCoordinateTransformContext &transformContext, const QgsProject *mProject );
+    QgsGeometryCheckContext( int precision, const QgsCoordinateReferenceSystem &mapCrs, const QgsCoordinateTransformContext &transformContext, const QgsProject *mProject = nullptr, const int uniqueIdFieldIndex = -1 );
```
Can you add a documentation note that uniqueIdFieldIndex was added in 4.0?

nyalldawson commented on src/analysis/vector/geometry_checker/qgsgeometrycheck.cpp:
```diff
@@ -175,3 +187,26 @@ double QgsGeometryCheck::scaleFactor( const QPointer<QgsVectorLayer> &layer ) co
+    return QgsGeometryCheck::Result::Success;
+  }
+
+  if ( !uniqueIds.contains( layerFeature.layerId() ) )
+  {
+    uniqueIds.insert( layerFeature.layerId(), QSet<QVariant>() );
+  }
+
+  const QVariant uniqueIdValue = layerFeature.feature().attribute( uniqueIdFieldIndex );
+  if ( uniqueIds[layerFeature.layerId()].contains( uniqueIdValue ) )
+  {
+    return QgsGeometryCheck::Result::DuplicatedUniqueId;
+  }
+  uniqueIds[layerFeature.layerId()].insert( uniqueIdValue );
```
Let's avoid some duplicate lookups:

```suggestion
  auto uniqueIdIt = uniqueIds.find( layerFeature.layerId() );
  if ( uniqueIdIt == uniqueIds.end() )
  {
    uniqueIdIt = uniqueIds.insert( layerFeature.layerId(), QSet<QVariant>() );
  }

  const QVariant uniqueIdValue = layerFeature.feature().attribute( uniqueIdFieldIndex );
  if ( uniqueIdIt.value().contains( uniqueIdValue ) )
  {
    return QgsGeometryCheck::Result::DuplicatedUniqueId;
  }
  uniqueIdIt.value().insert( uniqueIdValue );
```

alexbruy committed:
remove comment

alexbruy committed:
address review

alexbruy committed:
add documentation
